grpc-client-server/README.md

```diff
@@ -0,0 +1,188 @@
+
+Run your client against some website - how many repetitions do you need to see your client timeout?
+
+## Extra Challenges: Errors, Continuous Probing, Serve Prometheus Metrics
+
+These sections are optional - do them for an extra challenge if time permits.
+
+### Part 1:
+How do we know if the HTTP fetch succeeded at the server? Add a check to make sure it did.
+How should we deal with errors, i.e. if the endpoint isn't found?
+Modify your code and proto format to handle these cases.
+
+### Part 2: Run probes on a schedule
+Instead of getting probe times via the gRPC client we might consider running the endpoint probes on a schedule. Change the
+prober_server to accept a frequency argument instead of the old `repetitions` parameter, 
+and continue running probes every N seconds forever. 
+
+You will need to use a [goroutine](https://gobyexample.com/goroutines) to keep running probes in the background.
+
+Now, add a new client operation to get the value of the last probe run.
+To make this work, you will need to pass a value to your prober goroutine that it can use to store the most recent latency.
+You can store this value in your `server` struct.
+You will also need to use a `sync.Mutex` to guarantee that the goroutine that serves your client request sees the last value 
+written by the prober goroutine. A Tour of Go has a good example of [using a sync.Mutex](https://go.dev/tour/concurrency/9).
+
+
+### Part 3: Add Prometheus Metrics
+If you still have time, then let's learn something about how to monitor applications.
+
+In software operations, we want to know what our software is doing and how it is performing.
+One very useful technique is to have your program export metrics. Metrics are basically values that your 
+program makes available (via HTTP). Specialised programs, such as Prometheus, can then fetch metrics regularly
+from all the running instances of your program, store the history of these metrics, and do useful arithmetic on them
+(like computing rates, averages, and maximums). You can use this data to do troubleshooting and to alert if things 
+go wrong.
+
+Read the [Overview of Prometheus](https://prometheus.io/docs/introduction/overview/).
+
+Now add Prometheus metrics to your prober server. Every time you execute a probe, update a `gauge` metric that tracks the latency.
+Add a `label` specifying the endpoint being probed. 
+The [Prometheus Guide to Instrumenting a Go Application](https://prometheus.io/docs/guides/go-application/) has all the information you need to do this.
+
+Once you've run your program, use your client to execute periodic probes against some endpoint. 
+Now use the `curl` program or your browser to view the metrics on the port that you've chosen to serve them on.
+You should see a number of built-in Go metrics, plus your new gauge.
+
+If you use your client to start probing a second endpoint, you should see a second labelled metric appear.
+
+### Part 3: Scrape Prometheus Metrics
```

Comment:
Part 4

(We haven't really used the concept of "parts" so far. Not averse but I feel like part 1 is non negotiable, 2 is standalone, and 3+4 are a logical pair)